### PR TITLE
Rewrite For statement with CuPy

### PR DESCRIPTION
**Pull request by takagi**
Following up #7469.

### CODE REVIEW

File: cupyx/scipy/signal/_savitzky_golay.py [hunk 1]
```diff
@@ -22,9 +22,14 @@ def float_factorial(n: int) -> float:
...
-    for pv in p:
-        y = y * x + pv
+
+    # for pv in p:
+    #     y = y * x + pv
+    n = p.shape[0]
+    x = cupy.broadcast_to(x, (n - 1,) + x.shape)
+    x = cupy.cumprod(x, axis=0)[::-1]
+    p = cupy.expand_dims(p, axis=1)
+    y = cupy.sum(p[:-1] * x, axis=0) + p[-1]
```
Comment (emcastillo) [inline cupyx/scipy/signal/_savitzky_golay.py:32]: Can we compare performance?
I am not sure if this will be faster (probably yes)

Comment (takagi) [reply cupyx/scipy/signal/_savitzky_golay.py:32]: Yes, I'll have one.

**Comment by takagi** [comment]
Let me close this as I have no progress for now.

**Pull request closed without merging**